File: src/classifier/chargeTestSet.cpp
//! MIT License
//!
//! Copyright (c) 2025 Santos O. G., Helen C. S. C. Lima,
//! Copyright (c) 2009 Jr, C. N. S. Freitas
//! Permission is hereby granted, free of charge, to any person obtaining a copy
//! of this software and associated documentation files (the "Software"), to deal
//! in the Software without restriction, including without limitation the rights
//! to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//! copies of the Software, and to permit persons to whom the Software is
//! furnished to do so, subject to the following conditions:
//!
//! The above copyright notice and this permission notice shall be included in all
//! copies or substantial portions of the Software.

#include <iostream>
#include "chargeTestSet.h"
#include "utils.h"

using namespace std;

void ChargeTestSet::openTestFile() {
    fin.open(testFile.c_str());
    while (!fin) {
        DEBUG_ERR("\nErro na abertura do arquivo " << testFile << " em chargeTestSet.cpp!!!" << endl);
        exit(1);
    }
}

void ChargeTestSet::closeTestFile() {
    fin.close();
}

void ChargeTestSet::insertTestSet(const int &exampleId, const int &attributeId, const int &attributeValue) {
    testSet[exampleId][attributeId] = attributeValue;
}

int ChargeTestSet::getAttributeValueTestSet(const int &exampleId, const int &attributeId) {
    return testSet[exampleId][attributeId];
}

void ChargeTestSet::printTestSet() {
    int exampleId = 0;
    for (it_testSet = testSet.begin(); it_testSet != testSet.end(); it_testSet++) {
        for (it2_testSet = (*it_testSet).begin(); it2_testSet != (*it_testSet).end(); it2_testSet++) {
            cout << *it2_testSet << " ";
        }
        cout << getClassTestSet(exampleId++) << endl;
    }
}

void ChargeTestSet::insertClassTestSet(const int &exampleId, const string &attributeValue) {
    classTestSet[exampleId] = attributeValue;
}

string ChargeTestSet::getClassTestSet(const int &exampleId) {
    return classTestSet[exampleId];
}

void ChargeTestSet::printClassTestSet() {
    for (it_classTestSet = classTestSet.begin(); it_classTestSet != classTestSet.end(); it_classTestSet++) {
        cout << *it_classTestSet << endl;
    }
}

void ChargeTestSet::getTestSet() {
    string line;
    int i, begin, end, exampleId = 0, attributeId = 0;

    openTestFile();

    while (lowercase(line.substr(0, 5)) != "@data") {
        getline(fin, line);
    }

    while (getline(fin, line)) {
        if (line.substr(0, 1) == "%" || (int)line.size() == 0) {
            continue;
        }

        i = -1;
        while (i < (int)line.size()) {
            begin = i;
            if (begin == line.rfind(",")) {
                end = line.size();
                insertClassTestSet(exampleId, line.substr(begin + 1, end - (begin + 1)));
            } else {
                end = line.find(",", begin + 1);
                insertTestSet(exampleId, attributeId, str2int(line.substr(begin + 1, end - (begin + 1))));
            }
            i = end;
            attributeId++;
        }
        exampleId++;
        attributeId = 0;
    }

    closeTestFile();
}

File: src/charge_test_set.py
#! MIT License
#!
#! Copyright (c) 2025 Santos O. G., Helen C. S. C. Lima,
#! Permission is hereby granted, free of charge, to any person obtaining a copy
#! of this software and associated documentation files (the "Software"), to deal
#! in the Software without restriction, including without limitation the rights
#! to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#! copies of the Software, and to permit persons to whom the Software is
#! furnished to do so, subject to the following conditions:
#!
#! The above copyright notice and this permission notice shall be included in all
#! copies or substantial portions of the Software.

from utils import str2int
import sys

class ChargeTestSet:
    def __init__(self, test_file: str, number_of_test_examples: int, number_of_attributes: int):
        """
        Initializes the test set container.
        :param test_file: Path to the ARFF-style test file.
        :param number_of_test_examples: Expected number of test instances.
        :param number_of_attributes: Total attributes per instance (including class label).
        """
        self.test_file = test_file
        self.number_of_test_examples = number_of_test_examples
        self.number_of_attributes = number_of_attributes

        # allocate a 2D list for attribute values (exclude the class label)
        self.test_set = [
            [0] * (self.number_of_attributes - 1)
            for _ in range(self.number_of_test_examples)
        ]
        # allocate a list for class labels
        self.class_test_set = [''] * self.number_of_test_examples
        self._fin = None

    def open_test_file(self):
        """Opens the test file, exits on failure."""
        try:
            self._fin = open(self.test_file, 'r')
        except IOError:
            sys.stderr.write(f"[ERR] Error opening test file {self.test_file}\n")
            sys.exit(1)

    def close_test_file(self):
        """Closes the test file."""
        if self._fin:
            self._fin.close()
            self._fin = None

    def insert_test_set(self, example_id: int, attribute_id: int, attribute_value: int):
        """Inserts a single attribute value into the test_set matrix."""
        self.test_set[example_id][attribute_id] = attribute_value

    def get_attribute_value_test_set(self, example_id: int, attribute_id: int) -> int:
        """Retrieves an attribute value from the test_set matrix."""
        return self.test_set[example_id][attribute_id]

    def insert_class_test_set(self, example_id: int, class_value: str):
        """Inserts a class label for a given example index."""
        self.class_test_set[example_id] = class_value

    def get_class_test_set(self, example_id: int) -> str:
        """Retrieves the class label for a given example index."""
        return self.class_test_set[example_id]

    def get_test_set(self):
        """
        Reads the ARFF 'data' section and populates test_set and class_test_set.
        Skips comments (%) and blank lines. Uses lowercase and str2int from utils.
        """
        self.open_test_file()
        if self._fin is None:
            exit()

        # skip header until we hit '@data'
        for line in self._fin:
            if (line.strip()[:5]).lower() == '@data':
                break

        example_id = 0
        # read each data line
        for line in self._fin:
            line = line.strip()
            if not line or line.startswith('%'):
                continue

            parts = line.split(',')
            # parse attribute values
            for attr_id, raw_val in enumerate(parts[:-1]):
                self.test_set[example_id][attr_id] = str2int(raw_val)
            # parse class label
            self.class_test_set[example_id] = parts[-1]
            example_id += 1
        self.close_test_file()
